### ISSUE TITLE
Exception disk I/O error [778]

### ISSUE THREAD

**Issue by Fabi54321** [moneymanagerex/moneymanagerex#5318]
MMEX version:
 - [ ] 1.5.20
 - [ ] 1.5.21
 - [x] 1.6.0
 - [ ] 1.6.1 Beta
 - [ ] Other (please specify)

Note: bug reporters are expected to have verified the bug still exists
either in the last stable version of MMEX or on updated development code
(master branch).

Operating System:
 - [ ] Windows
 - [x] Mac OSX (Ventura)
 - [ ] Linux 

# Description of the bug

After upgrade from 1.5.21 to 1.6.0 and a DB optimization (which was suggested), I get errors like below when I switch between account views:
<img width="297" alt="image" src="https://user-images.githubusercontent.com/13711459/200647617-6a0375f7-c93f-4928-a290-f07d958ac2b8.png">
<img width="297" alt="image" src="https://user-images.githubusercontent.com/13711459/200647702-e0df33df-c4b4-416e-ae4a-23f1fbcba856.png">

When closing MMEX:
![image](https://user-images.githubusercontent.com/13711459/200650269-84a4a520-4c6a-44e2-ac87-e619b7eeef1d.png)

It also seems that MMEX crashes after that message because reopening shows:
<img width="297" alt="image" src="https://user-images.githubusercontent.com/13711459/200650738-393b87e8-cba2-491d-bcdf-8d680d0885df.png">

# Reproduction

Is the bug reproducible? 
 - [ ] Always 
 - [x] Randomly 
 - [ ] Happened only once

Reproduction steps:
Unclear. It happens only in some sessions but persists in that cases until MMEX is restarted.

Expected result:
No error should occur.

Actual result:
Error described above is visible. Consequences of error unclear.

# Additional information
I also have the issue that text in left pane is blurry/bold until it's clicked or scrolled. Scrolling "fixes" the whole list:
<img width="144" alt="image" src="https://user-images.githubusercontent.com/13711459/200652103-515e9238-d84a-46f8-91b1-7731fbfbe445.png">
<img width="151" alt="image" src="https://user-images.githubusercontent.com/13711459/200652134-ae8e8014-4a1a-4a41-b546-ec6ee0e6e355.png">
Should I create another ticket for that?

![image](https://user-images.githubusercontent.com/13711459/200648795-2b898204-90af-466c-b8f6-8f6cb549a264.png)

**Comment by Fabi54321** [comment]
I also just noticed several new backup files created with versioning. Is it supposed to be like that?
<img width="274" alt="image" src="https://user-images.githubusercontent.com/13711459/200653606-c185845d-8c80-43bb-b0e9-64520f703956.png">

**Comment by tactilis** [comment]
> I also just noticed several new backup files created with versioning. Is it supposed to be like that?

The backup files are because version 1.6.0 needs to upgrade the database structure from DB v7 to DB v15 so that some additional MMEX capabilities can be supported.

It does this in a series of steps and keeps a backup file as each change is made to the .mmb file.

_upgrade_V7 is a copy of your your pre-1.6.0 database. So you can take a copy of this if you need to go back to using 1.5.20 or 1.5.21

**Comment by whalley** [comment]
The error looks like as if it is a file permissions issue. Please check the file permissions on the DB files.
Also check these settings. I assume your DB file is one of these directories?
![CleanShot 2022-11-08 at 20 03 02](https://user-images.githubusercontent.com/487345/200663716-c00aad96-c00e-4fc9-914f-28e0ce148d56.png)

Regarding the blurry text. I am aware of this and encounter it on one of my monitors. Think it may be related to display driver interaction. I had similar following an O/S upgrade before and it went a month or so later - may be the same following Ventura update. I will be monitoring this. Feel free to raise issue to track.

**Comment by Fabi54321** [comment]
> The backup files are because version 1.6.0 needs to upgrade the database structure from DB v7 to DB v15 so that some additional MMEX capabilities can be supported.
> 
> It does this in a series of steps and keeps a backup file as each change is made to the .mmb file.
> 
> _upgrade_v7 is a copy of your your pre-1.6.0 database. So you can take a copy of this if you need to go back to using 1.5.20 or 1.5.21

@tactilis Thanks for the detailed explanation :) I already copied the whole folder to a backup location to be on the safe side ;)

**Comment by Fabi54321** [comment]
> The error looks like as if it is a file permissions issue. Please check the file permissions on the DB files. Also check these settings. I assume your DB file is one of these directories?

Thanks @whalley. Yes, my database is stored in OneDrive. I checked the permissions and **mmex already has permission** to access OneDrive:
![image](https://user-images.githubusercontent.com/13711459/200678256-b4f129cb-e2e6-414f-a5df-ce1404c86758.png)
Wouldn't a file permissions issue cause the error to be always there, not only sometimes?

> Regarding the blurry text. I am aware of this and encounter it on one of my monitors. Think it may be related to display driver interaction. I had similar following an O/S upgrade before and it went a month or so later - may be the same following Ventura update. I will be monitoring this. Feel free to raise issue to track.

Thanks, then I'll just wait some time. The blurry text is not a big issue :)

**Comment by tactilis** [comment]
> Yes, my database is stored in OneDrive. I checked the permissions and mmex already has permission to access OneDrive
> Wouldn't a file permissions issue cause the error to be always there, not only sometimes?

I suggest that you copy your <database name.mmb>_upgrade_v7_2022-11-02.bak to TESTING.mmb in a location other than OneDrive and then try opening it with MMEX 1.6.0.  This will do the database upgrade to DB v15 (and create the TESTING.mmb_upgrade\* files).

Does 1.6.0 now work correctly with the upgraded TESTING.mmb database?

**Comment by Fabi54321** [comment]
> > Yes, my database is stored in OneDrive. I checked the permissions and mmex already has permission to access OneDrive
> > Wouldn't a file permissions issue cause the error to be always there, not only sometimes?
> 
> I suggest that you copy your _upgrade_v7_2022-11-02.bak to TESTING.mmb in a location other than OneDrive and then try opening it with MMEX 1.6.0. This will do the database upgrade to DB v15 (and create the TESTING.mmb_upgrade* files).
> 
> Does 1.6.0 now work correctly with the upgraded TESTING.mmb database?

Currently I was not even able to reproduce it any longer with the original updated DB file located in OneDrive. (tried at least 5 times).
Anyhow I did what you suggested but also with testing copy it worked fine (also tried at least 5 times).

Is there a way I can make sure that my database has no issue before continuing productive?

**Comment by whalley** [comment]
> Thanks @whalley. Yes, my database is stored in OneDrive. I checked the permissions and **mmex already has permission** to access OneDrive.
Wouldn't a file permissions issue cause the error to be always there, not only sometimes?

Sorry I missed that it only occurs sometimes. Not sure what is happening, how often does it occur? Do other database tables get updated OK, e.g. if you add/change a transaction?

**Comment by Dandy64** [comment]
Just FYI. I have been using Money Manager Ex 1.6.0 portable 64.

No problem happened after updating the DB.

**Comment by Fabi54321** [comment]
> Sorry I missed that it only occurs sometimes. Not sure what is happening, how often does it occur? Do other database tables get updated OK, e.g. if you add/change a transaction?

No problem. Originally it happened approx. every second session (from opening mmex until closing it again) and throughout the session it was either 100% reproducible or not at all. 
Unfortunately it currently does not reproduce at all. I tried to force it to fail by removing the file permissions in OS X when open, but this only comes to effect when closing mmex and results in a completely different error.

I am wondering what else I can do to reproduce or if it is safe to continue without a solution...

**Comment by Fabi54321** [comment]
> Just FYI. I have been using Money Manager Ex 1.6.0 portable 64.
> 
> No problem happened after updating the DB.

Good point - upgrade of the DB was also successful on my side according to message box. Just after upgrade and (also successful) DB optimization the errors occurred.

**Comment by whalley** [comment]
> I am wondering what else I can do to reproduce or if it is safe to continue without a solution...

Just continue... Not sure what happened, nobody else has reported any issues. 

Closing this.

**Comment by whalley** [comment]
> > Regarding the blurry text. I am aware of this and encounter it on one of my monitors. Think it may be related to display driver interaction. I had similar following an O/S upgrade before and it went a month or so later - may be the same following Ventura update. I will be monitoring this. Feel free to raise issue to track.
> 
> Thanks, then I'll just wait some time. The blurry text is not a big issue :)

Looks to be a wxWidgets issue... I'll raise an issue for MMEX to track this.
https://github.com/wxWidgets/wxWidgets/issues/22629
